File: spikes/1.2-wasm-embedding/hosts/jvm/src/main/java/io/pact/janus/spike/Bench.java
package io.pact.janus.spike;

import com.dylibso.chicory.compiler.MachineFactoryCompiler;
import com.dylibso.chicory.runtime.ExportFunction;
import com.dylibso.chicory.runtime.Instance;
import com.dylibso.chicory.wasm.Parser;
import com.dylibso.chicory.wasm.WasmModule;

import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.Arrays;
import java.util.LinkedHashMap;
import java.util.Map;
import java.util.function.Function;
import java.util.function.Supplier;

/**
 * JVM host leg: Chicory 1.4 (pure-JVM runtime, zero native deps).
 * Attempts the component binary (recording the failure), then benchmarks the
 * core module + C-ABI shim under both the interpreter and the runtime
 * bytecode compiler backends.
 */
public final class Bench {

    interface Pipe {
        byte[] call(byte[] request);
    }

    static Pipe makePipe(Instance instance) {
        ExportFunction alloc = instance.export("alloc");
        ExportFunction dealloc = instance.export("dealloc");
        ExportFunction call = instance.export("call");
        return request -> {
            int reqLen = request.length;
            int reqPtr = (int) alloc.apply(reqLen)[0];
            instance.memory().write(reqPtr, request);
            long packed = call.apply(reqPtr, reqLen)[0];
            int respPtr = (int) (packed >>> 32);
            int respLen = (int) (packed & 0xffffffffL);
            byte[] response = instance.memory().readBytes(respPtr, respLen);
            dealloc.apply(respPtr, respLen);
            dealloc.apply(reqPtr, reqLen);
            return response;
        };
    }

    static void stats(String name, long[] samples) {
        long[] s = samples.clone();
        Arrays.sort(s);
        Function<Double, Double> at = q -> s[Math.min(s.length - 1, (int) (q * s.length))] / 1e3;
        System.out.printf("%s: median %.1fµs  p95 %.1fµs  min %.1fµs%n",
                name, at.apply(0.5), at.apply(0.95), s[0] / 1e3);
    }

    static void bench(String name, Runnable fn, int warmup, int iters) {
        for (int i = 0; i < warmup; i++) fn.run();
        long[] samples = new long[iters];
        for (int i = 0; i < iters; i++) {
            long t0 = System.nanoTime();
            fn.run();
            samples[i] = System.nanoTime() - t0;
        }
        stats(name, samples);
    }

    static void check(String name, byte[] resp) {
        String s = new String(resp, StandardCharsets.UTF_8);
        if (!s.startsWith("{\"ok\"")) {
            throw new IllegalStateException(name + ": unexpected response " + s.substring(0, Math.min(200, s.length())));
        }
    }

    static void runLeg(String title, Supplier<Pipe> makeFresh, Map<String, byte[]> frames) {
        System.out.println("--- " + title);
        long[] cold = new long[20];
        Pipe pipe = null;
        for (int i = 0; i < cold.length; i++) {
            long t0 = System.nanoTime();
            pipe = makeFresh.get();
            cold[i] = System.nanoTime() - t0;
        }
        System.out.printf("cold-start: first %.1fms%n", cold[0] / 1e6);
        stats("cold-start (per-iter)", cold);

        check("handshake", pipe.call(frames.get("handshake")));
        check("echo-small", pipe.call(frames.get("echo-small")));
        check("match-100k", pipe.call(frames.get("match-100k")));

        Pipe p = pipe;
        bench("echo-small", () -> p.call(frames.get("echo-small")), 500, 5000);
        bench("echo-100k", () -> p.call(frames.get("echo-100k")), 50, 500);
        bench("match-100k", () -> p.call(frames.get("match-100k")), 20, 200);
    }

    public static void main(String[] args) throws Exception {
        Path base = Path.of("../..").toAbsolutePath().normalize();
        byte[] doc = Files.readAllBytes(base.resolve("payloads/order-100kb.json"));
        String docStr = new String(doc, StandardCharsets.UTF_8);

        Map<String, byte[]> frames = new LinkedHashMap<>();
        frames.put("handshake", "{\"op\":\"handshake\",\"protocol-versions\":[1]}".getBytes(StandardCharsets.UTF_8));
        frames.put("echo-small", "{\"op\":\"echo\",\"payload\":{\"ping\":1}}".getBytes(StandardCharsets.UTF_8));
        frames.put("echo-100k", ("{\"op\":\"echo\",\"payload\":" + docStr + "}").getBytes(StandardCharsets.UTF_8));
        frames.put("match-100k", ("{\"op\":\"match-type\",\"expected\":" + docStr + ",\"actual\":" + docStr + "}")
                .getBytes(StandardCharsets.UTF_8));

        // Attempt 1: the component binary — record what Chicory does with it.
        try {
            Parser.parse(base.resolve("engine-toy/target/wasm32-wasip1/release/toy_engine_component.wasm"));
            System.out.println("component binary: parsed (unexpected!)");
        } catch (Exception e) {
            System.out.println("component binary: " + e.getClass().getSimpleName() + ": " + e.getMessage());
        }

        byte[] coreBytes = Files.readAllBytes(
                base.resolve("engine-toy/target/wasm32-unknown-unknown/release/toy_engine_core.wasm"));

        runLeg("interpreter", () -> {
            WasmModule module = Parser.parse(coreBytes);
            return makePipe(Instance.builder(module).build());
        }, frames);

        runLeg("bytecode compiler", () -> {
            WasmModule module = Parser.parse(coreBytes);
            return makePipe(Instance.builder(module)
                    .withMachineFactory(MachineFactoryCompiler::compile)
                    .build());
        }, frames);
    }
}

File: spikes/1.2-wasm-embedding/hosts/python/bench.py
# Python host leg: wasmtime-py 48.
# Path A: the real component via wasmtime.component (dynamic API, add_wasip2 for WASI).
# Path B: core module + hand-built C-ABI shim via the classic API.
import json
import statistics
import time
from pathlib import Path

import wasmtime
import wasmtime.component as wc

HERE = Path(__file__).parent
COMPONENT = (HERE / "../../engine-toy/target/wasm32-wasip1/release/toy_engine_component.wasm").resolve()
CORE = (HERE / "../../engine-toy/target/wasm32-unknown-unknown/release/toy_engine_core.wasm").resolve()
DOC = (HERE / "../../payloads/order-100kb.json").resolve().read_text()

FRAMES = {
    "handshake": b'{"op":"handshake","protocol-versions":[1]}',
    "echo-small": b'{"op":"echo","payload":{"ping":1}}',
    "echo-100k": ('{"op":"echo","payload":%s}' % DOC).encode(),
    "match-100k": ('{"op":"match-type","expected":%s,"actual":%s}' % (DOC, DOC)).encode(),
}


def report(name, samples_ns):
    s = sorted(samples_ns)
    at = lambda q: s[min(len(s) - 1, int(q * len(s)))] / 1e3
    print(f"{name}: median {at(0.5):.1f}µs  p95 {at(0.95):.1f}µs  min {s[0]/1e3:.1f}µs")


def bench(name, fn, warmup, iters):
    for _ in range(warmup):
        fn()
    samples = []
    for _ in range(iters):
        t0 = time.perf_counter_ns()
        fn()
        samples.append(time.perf_counter_ns() - t0)
    report(name, samples)


def check(name, resp_bytes, expect_key="ok"):
    resp = json.loads(bytes(resp_bytes))
    assert expect_key in resp, f"{name}: unexpected response {str(resp)[:200]}"


def run_leg(title, make_pipe):
    print(f"--- {title}")
    cold, pipe = [], None
    for _ in range(20):
        t0 = time.perf_counter_ns()
        pipe = make_pipe()
        cold.append(time.perf_counter_ns() - t0)
    print(f"cold-start: first {cold[0]/1e6:.1f}ms  median {statistics.median(cold)/1e6:.1f}ms")
    check("handshake", pipe(FRAMES["handshake"]))
    check("echo-small", pipe(FRAMES["echo-small"]))
    check("match-100k", pipe(FRAMES["match-100k"]))
    bench("echo-small", lambda: pipe(FRAMES["echo-small"]), 500, 5000)
    bench("echo-100k", lambda: pipe(FRAMES["echo-100k"]), 50, 500)
    bench("match-100k", lambda: pipe(FRAMES["match-100k"]), 20, 200)


# --- Path A: component model
component_bytes = COMPONENT.read_bytes()


def make_component_pipe():
    engine = wasmtime.Engine()
    store = wasmtime.Store(engine)
    component = wc.Component(engine, component_bytes)
    linker = wc.Linker(engine)
    linker.add_wasip2()
    store.set_wasi(wasmtime.WasiConfig())
    instance = linker.instantiate(store, component)
    idx = instance.get_export_index(store, "pact:toy-engine/pipe@0.1.0")
    fidx = instance.get_export_index(store, "call", idx)
    func = instance.get_func(store, fidx)
    return lambda req: func(store, req)


# --- Path B: core + shim
core_bytes = CORE.read_bytes()


def make_core_pipe():
    engine = wasmtime.Engine()
    store = wasmtime.Store(engine)
    module = wasmtime.Module(engine, core_bytes)
    instance = wasmtime.Instance(store, module, [])
    exports = instance.exports(store)
    memory, alloc, dealloc, call = exports["memory"], exports["alloc"], exports["dealloc"], exports["call"]

    def pipe(request):
        req_len = len(request)
        req_ptr = alloc(store, req_len)
        memory.write(store, request, req_ptr)
        packed = call(store, req_ptr, req_len)
        resp_ptr, resp_len = (packed >> 32) & 0xFFFFFFFF, packed & 0xFFFFFFFF
        response = memory.read(store, resp_ptr, resp_ptr + resp_len)
        dealloc(store, resp_ptr, resp_len)
        dealloc(store, req_ptr, req_len)
        return response

    return pipe


if __name__ == "__main__":
    run_leg("component (wasmtime.component, dynamic)", make_component_pipe)
    run_leg("core + C-ABI shim", make_core_pipe)
